### PR TITLE
network: Add support to emit and receive log server information

### PR DESCRIPTION
**Pull request by ssahani**
Implements DHCP option 7

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp-lease.c [hunk 1]
```diff
@@ -129,6 +129,17 @@ int sd_dhcp_lease_get_sip(sd_dhcp_lease *lease, const struct in_addr **addr) {
         return (int) lease->sip_size;
 }
 
+int sd_dhcp_lease_get_log(sd_dhcp_lease *lease, const struct in_addr **addr) {
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-lease.c]: `log` is quite general term. Maybe, `sd_dhcp_lease_get_log_server()` is better.

File: src/libsystemd-network/dhcp-lease-internal.h [hunk 2]
```diff
@@ -61,6 +61,9 @@ struct sd_dhcp_lease {
         struct in_addr *sip;
         size_t sip_size;
 
+        struct in_addr *log;
+        size_t log_size;
```
Comment (yuwata) [inline src/libsystemd-network/dhcp-lease-internal.h]: I prefer log_server(s), and log_server_size.

File: src/libsystemd-network/sd-dhcp-lease.c [hunk 3]
```diff
@@ -601,6 +613,12 @@ int dhcp_lease_parse_options(uint8_t code, uint8_t len, const void *option, void
                         log_debug_errno(r, "Failed to parse SIP server, ignoring: %m");
                 break;
 
+     case SD_DHCP_OPTION_LOG_SERVER:
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-lease.c]: Indentation?

File: src/libsystemd-network/sd-dhcp-server.c [hunk 4]
```diff
@@ -1181,6 +1191,32 @@ int sd_dhcp_server_set_sip(sd_dhcp_server *server, const struct in_addr sip[], u
...
+                server->n_log = 0;
+        } else {
+                struct in_addr *c;
+
+                c = newdup(struct in_addr, log, n);
+                if (!c)
+                        return -ENOMEM;
+
+                free(server->log);
+                server->log = c;
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: `free_and_replace()`?

File: src/libsystemd-network/sd-dhcp-lease.c [hunk 5]
```diff
@@ -1066,6 +1085,7 @@ int dhcp_lease_load(sd_dhcp_lease **ret, const char *lease_file) {
                            "DNS", &dns,
                            "NTP", &ntp,
                            "SIP", &sip,
+                           "LOG", &log,
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-lease.c]: I prefer LOG_SERVER or LOG_SERVERS.

File: src/network/networkd-dhcp-common.c [hunk 6]
```diff
@@ -126,6 +126,38 @@ int config_parse_dhcp_use_sip(
         return 0;
 }
 
+int config_parse_dhcp_use_log(
```
Comment (yuwata) [inline src/network/networkd-dhcp-common.c]: It seems not necessary to introduce own config parser but please use `config_parse_bool()`.

File: src/network/networkd-dhcp-server.c [hunk 7]
```diff
@@ -140,6 +140,55 @@ static int link_push_uplink_ntp_to_dhcp_server(Link *link, sd_dhcp_server *s) {
...
+                addresses[n_addresses++] = ia.in;
+        }
+
+        if (link->network->dhcp_use_log && link->dhcp_lease) {
+                const struct in_addr *da = NULL;
+                int j, n;
+
+                n = sd_dhcp_lease_get_log(link->dhcp_lease, &da);
+                if (n > 0) {
+
```
Comment (yuwata) [inline src/network/networkd-dhcp-server.c:176]: Unnecessary empty line.

File: man/systemd.network.xml [hunk 8]
```diff
@@ -1903,6 +1903,16 @@
         <varname>DNS=</varname>.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>LogServer=</varname></term>
```
Comment (keszybz) [inline man/systemd.network.xml]: This takes multiple items, right? So "LogServers=".

File: man/systemd.network.xml [hunk 9]
```diff
@@ -1903,6 +1903,16 @@
...
+
+        <listitem><para>Similar to the <varname>DNS=</varname> settings
+        described above, these settings configure whether and what log server
```
Comment (keszybz) [inline man/systemd.network.xml]: Hmm, maybe add a short explanation what "log server" is? While "DNS" and "POP3" and "SMTP" are all well known terms, "log" is rather obscure.

File: src/network/networkd-network-gperf.gperf [hunk 10]
```diff
@@ -207,6 +207,7 @@ DHCPServer.EmitNTP,                          config_parse_bool,
 DHCPServer.NTP,                              config_parse_dhcp_server_ntp,                             0,                             0
 DHCPServer.EmitSIP,                          config_parse_bool,                                        0,                             offsetof(Network, dhcp_server_emit_sip)
 DHCPServer.SIP,                              config_parse_dhcp_server_sip,                             0,                             0
+DHCPServer.LogServer,                        config_parse_dhcp_server_log_servers,                     0,                             0
```
Comment (keszybz) [inline src/network/networkd-network-gperf.gperf]: LogServers

File: src/network/networkd-network.h [hunk 11]
```diff
@@ -296,6 +299,7 @@ struct Network {
 
         char **ntp;
         char **sip;
+        char **log_server;
```
Comment (keszybz) [inline src/network/networkd-network.h]: log_servers

File: src/libsystemd/sd-network/sd-network.c [hunk 12]
```diff
@@ -256,6 +256,10 @@ _public_ int sd_network_link_get_sip_servers(int ifindex, char ***ret) {
         return network_link_get_strv(ifindex, "SIP", ret);
 }
 
+_public_ int sd_network_link_get_log_servers(int ifindex, char ***ret) {
```
Comment (keszybz) [inline src/libsystemd/sd-network/sd-network.c]: Please make parameter naming with the prototype.

File: src/network/networkctl.c [hunk 13]
```diff
@@ -1226,7 +1226,8 @@ static int link_status_one(
                 sd_hwdb *hwdb,
                 const LinkInfo *info) {
 
-        _cleanup_strv_free_ char **dns = NULL, **ntp = NULL, **sip = NULL, **search_domains = NULL, **route_domains = NULL;
+        _cleanup_strv_free_ char **dns = NULL, **ntp = NULL, **sip = NULL, **search_domains = NULL, **route_domains = NULL,
+                **log_server = NULL;
```
Comment (keszybz) [inline src/network/networkctl.c]: log_servers

File: man/systemd.network.xml [hunk 14]
```diff
@@ -1903,6 +1903,16 @@
         <varname>DNS=</varname>.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>LogServers</varname></term>
```
Comment (keszybz) [inline man/systemd.network.xml]: LogServers=

File: man/systemd.network.xml [hunk 15]
```diff
@@ -1903,6 +1903,16 @@
...
+
+        <listitem><para>Specifies a list of MIT-LCS UDP servers available to the client.
```
Comment (keszybz) [inline man/systemd.network.xml:1909]: Dunno, it's still not obvious what this is used for based on this description, but OK.

**Comment by poettering** [comment]
I m sorry, but I do not see the point of this. The logging protocol this is for is not syslog, but "MIT-LCS", which noone uses or knows anymore.

Let's not do this.

See:

https://marc.info/?l=dhcp-server&m=106807935712328
https://marc.info/?l=dhcp-server&m=106814567313604&w=2

**Comment by ssahani** [comment]
Sorry for the confusion I thought this syslog server https://www.incognito.com/tutorials/dhcp-options-in-plain-english/
https://quickview.cloudapps.cisco.com/quickview/bug/CSCum57354
https://books.google.co.in/books?id=EFoPDAAAQBAJ&pg=PT640&lpg=PT640&dq=dhcp+syslog+option+7&source=bl&ots=UYufFqMJOZ&sig=ACfU3U0WEX4z-D3Lj0Z2Hkzn1fEnDGJjRQ&hl=en&sa=X&ved=2ahUKEwiVgr3y1MPoAhUQzTgGHYjKBLcQ6AEwEHoECAoQAQ#v=onepage&q=dhcp%20syslog%20option%207&f=false

I think later they renamed to syslog server. But No thing strict we have other options to add  other others 
options if someone asks. I just have to replace the variables.

**Comment by poettering** [comment]
Do you know any software actually using the field for syslog logging?

**Comment by ssahani** [comment]
No I don'ts know.  It just a way to provide the clients  the centralised syslog server .

**Pull request closed without merging**